pkg/kvstore/lock.go

```diff
@@ -38,22 +38,28 @@ type Lock struct {
 }
 
 var (
-	lockPathsMU lock.Mutex
+	lockPathsMU lock.RWMutex
 	lockPaths   = map[string]*lock.Mutex{}
 )
 
 // LockPath locks the specified path and returns the Lock
 func LockPath(path string) (*Lock, error) {
+	var (
+		localLock *lock.Mutex
+		ok        bool
+	)
 	lockPathsMU.Lock()
-	if lockPaths[path] == nil {
-		lockPaths[path] = &lock.Mutex{}
+	localLock, ok = lockPaths[path]
+	if !ok {
+		localLock = &lock.Mutex{}
+		lockPaths[path] = localLock
 	}
 	lockPathsMU.Unlock()
 
 	trace("Creating lock", nil, logrus.Fields{fieldKey: path})
 
 	// Take the local lock as both etcd and consul protect per client
-	lockPaths[path].Lock()
+	localLock.Lock()
 
 	lock, err := Client().LockPath(path)
 	if err != nil {
```

Comment:
we need to use localLock in the line below this (which I can't comment on in github :/ ) when we release it in the error case.

Reply:
Correct, thanks! Fixed